[Docs] Distinguish warm vs cold vs frozen tiers (#84155)
elastic/elasticsearch: docs/reference/datatiers.asciidoc

and hold your most recent, most-frequently-accessed data.
* <<warm-tier, Warm tier>> nodes hold time series data that is accessed less-frequently
and rarely needs to be updated.

> * <<cold-tier, Cold tier>> nodes hold time series data that is accessed 
> infrequently and not normally updated. <<fully-mounted,Fully mounted
> indices>> of <<ilm-searchable-snapshot,{search-snaps}>> may be kept on the
> cold tier, eliminating the need for replicas with approximately 50% storage
> reduction compared to the warm tier.

Author: ```suggestion
* <<cold-tier,Cold tier>> nodes hold time series data that is accessed
infrequently and not normally updated. To save space, you can keep
<<fully-mounted,fully mounted indices>> of
<<ilm-searchable-snapshot,{search-snaps}>> on the cold tier. These fully mounted
indices eliminate the need for replicas, reducing required disk space by
approximately 50% compared to the regular indices.
+
Alternatively, you can use the cold tier to store regular indices with replicas instead
of using {search-snaps}. This lets you store older data on less expensive hardware
but doesn't reduce required disk space compared to the warm tier.
```

and hold your most recent, most-frequently-accessed data.
* <<warm-tier, Warm tier>> nodes hold time series data that is accessed less-frequently
and rarely needs to be updated.

> * <<cold-tier, Cold tier>> nodes hold time series data that is accessed infrequently and not normally updated. <<fully-mounted,Fully mounted indices>> of <<ilm-searchable-snapshot,{search-snaps}>> are kept on cold tier eliminating the need for replicas with approximately 50% storage reduction compares to warm tiers.

Reviewer: I think this implies that it happens automatically, so I would recommend something like:
```suggestion
* <<cold-tier, Cold tier>> nodes hold time series data that is accessed infrequently and not normally updated. <<fully-mounted,Fully mounted indices>> of <<ilm-searchable-snapshot,{search-snaps}>> may be kept on the cold tier, eliminating the need for replicas with approximately 50% storage reduction compared to the warm tier.
```

> * <<cold-tier, Cold tier>> nodes hold time series data that is accessed infrequently and not normally updated. <<fully-mounted,Fully mounted indices>> of <<ilm-searchable-snapshot,{search-snaps}>> may be kept on the cold tier, eliminating the need for replicas with approximately 50% storage reduction compared to the warm tier.

Reviewer 2: Two suggestions:

- I'd point out that the storage savings is in disk space.
- I'd point out that you can use regular indices on the cold tier if wanted.

```suggestion
* <<cold-tier,Cold tier>> nodes hold time series data that is accessed
infrequently and not normally updated. To save space, you can keep
<<fully-mounted,fully mounted indices>> of
<<ilm-searchable-snapshot,{search-snaps}>> on the cold tier. These fully mounted
indices eliminate the need for replicas, reducing required disk space by
approximately 50% compared to the regular indices.
+
Alternatively, you can the cold tier to store regular indices with replicas instead
of using {search-snaps}. This lets you store older data on less expensive hardware
but doesn't reduce required disk space compared to the warm tier.
```

Author: 🤦 , I somehow missed this suggestion while I was working my way through the above.  I have added this commit in https://github.com/elastic/elasticsearch/pull/84155#discussion_r813521447 and in the main body.
 
To avoid duplication, I'd prefer to simplify the cold tier description in the summary at the top, and put the detailed description in the content, but I felt all of these points are important though, um...

Additionally, I am reading https://www.elastic.co/guide/en/elasticsearch/reference/current/modules-node.html and felt that some of the descriptions around cold or frozen data **nodes** can be updated accordingly or linked to content on this page.  Can you assist with that please?

> * <<cold-tier, Cold tier>> nodes hold time series data that is accessed infrequently and not normally updated. <<fully-mounted,Fully mounted indices>> of <<ilm-searchable-snapshot,{search-snaps}>> may be kept on the cold tier, eliminating the need for replicas with approximately 50% storage reduction compared to the warm tier.

Author: Great suggestion. Could you check if my latest commit satisfies the above?

I made quite a lot of modifications in the cold tier paragraph. Mainly to point out that one can use cold tier just like warm tier, or use searchable snapshot. 

```
Cold tier can be used for storing rarely updated time series data similar to 
the warm tier.  It can also be used for storing read-only
<<fully-mounted,fully mounted indices>> of
<<ilm-searchable-snapshot,{search-snaps}>>.  When one chooses to use searchable
snapshot on the cold tier, it eliminates the need for replicas with
approximately 50% storage reduction compared to the warm tier.
```

This is indeed a frequent question users are confused about the cold tier.

> * <<frozen-tier, Frozen tier>> nodes hold time series data that is accessed rarely and never updated. <<partially-mounted,Partially mounted indices>> of <<ilm-searchable-snapshot,{search-snaps}>> are kept on frozen tier extend the storage capacity even further by 20 times compares to warm tiers.

Reviewer: ```suggestion
* <<frozen-tier, Frozen tier>> nodes hold time series data that is accessed rarely and never updated. <<partially-mounted,Partially mounted indices>> of <<ilm-searchable-snapshot,{search-snaps}>> are kept on frozen tier to extend the storage capacity even further by up to 20 times compare to the warm tier.
```

Author: One thing I never grasp with English grammar is `tense`.  

In the previous [commit](https://github.com/elastic/elasticsearch/commit/5c97e845db636d3df4dd116c5519b4a2c0512e73#r809549540), you propose to change `compares` to `compared`.  In this sentence, you propose to change `compares` to `compare`.  

What have I missed? :P

Reviewer: I had a typo, they should both be `compared`, I fixed my typo now :)

> * <<frozen-tier, Frozen tier>> nodes hold time series data that is accessed rarely and never updated. <<partially-mounted,Partially mounted indices>> of <<ilm-searchable-snapshot,{search-snaps}>> are kept on frozen tier to extend the storage capacity even further by up to 20 times compare to the warm tier.

Author: I have modified the paragraph and used the word `exclusively`, hope that can justify explicitly mention?

```
* <<frozen-tier, Frozen tier>> nodes hold time series data that is accessed 
rarely and never updated. It is necessary to have a snapshot repository setup
in order to use frozen tier. The frozen tier stores
<<partially-mounted,Partially mounted indices>> of
<<ilm-searchable-snapshot,{search-snaps}>> exclusively.  This extends the
storage capacity even further by up to 20 times compare to the warm tier. 
```

Additionally, I have also seen quite a lot of confusion around frozen node setup where users may not realize it depends on a snapshot repository. So I added the sentence to emphasize the need.

[[cold-tier]]
=== Cold tier

> Cold tier can be used for storing rarely updated time series data similar to 

Author: ```suggestion
Cold tier can be used for storing infrequently updated time series data similar to 
```
Change to infrequently to be consistent with other places.

> the warm tier.  It can also be used to for storing read-only

Author: ```suggestion
the warm tier.  It can also be used for storing read-only
```

Once data is no longer being queried, or being queried rarely, it may move from
the cold tier to the frozen tier where it stays for the rest of its life.

> It is necessary to have a snapshot repository setup in order to use frozen tier.

Reviewer 2: Nit: rewording
```suggestion
The frozen tier requires a snapshot repository.
```
